base/gfspclient/gater.go

```diff
@@ -208,6 +224,137 @@ func (s *GfSpClient) NotifyDestSPMigrateSwapOut(ctx context.Context, destEndpoin
+	}
+
+	signedMsg, err := hex.DecodeString(resp.Header.Get(GnfdSignedApprovalMsgHeader))
+	if err != nil {
+		return types.GfSpQuotaInfo{}, err
+	}
+
+	quotaResult := types.GfSpQuotaInfo{}
+	if err = proto.Unmarshal(signedMsg, &quotaResult); err != nil {
+		return types.GfSpQuotaInfo{}, err
+	}
+
+	return quotaResult, nil
+}
+
+// PostMigrateBucket is used to notify src sp the completion of bucket migrate before dest sp send CompleteMigrateBucket Tx
+func (s *GfSpClient) PostMigrateBucket(ctx context.Context, srcSPEndpoint string, postMsg *types.GfSpPostMigrateBucket) (types.GfSpQuotaInfo, error) {
+	bucketID := postMsg.GetBucketId()
+	req, err := http.NewRequest(http.MethodGet, srcSPEndpoint+PostMigrateBucketPath, nil)
+	if err != nil {
+		log.CtxErrorw(ctx, "client failed to connect to gateway", "endpoint", srcSPEndpoint, "error", err)
+		return types.GfSpQuotaInfo{}, err
+	}
+
+	msg, err := json.Marshal(postMsg)
+	if err != nil {
+		log.CtxErrorw(ctx, "failed to post migrate bucket to src sp due to marshal error", "error", err)
+		return types.GfSpQuotaInfo{}, err
+	}
+	req.Header.Add(GnfdPostMigrateBucketMsgHeader, hex.EncodeToString(msg))
+	resp, err := s.HTTPClient(ctx).Do(req)
+	if err != nil {
+		log.Errorw("failed to post migrate bucket to src sp", "bucket_id", bucketID, "error", err)
+		return types.GfSpQuotaInfo{}, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return types.GfSpQuotaInfo{}, fmt.Errorf("failed to post migrate bucket to src sp, bucket(%d), status_code(%d), endpoint(%s)", bucketID, resp.StatusCode, srcSPEndpoint)
+	}
+
+	signedMsg, err := hex.DecodeString(resp.Header.Get(GnfdSignedApprovalMsgHeader))
+	if err != nil {
+		return types.GfSpQuotaInfo{}, err
+	}
+
+	quotaResult := types.GfSpQuotaInfo{}
+	if err = proto.Unmarshal(signedMsg, &quotaResult); err != nil {
+		return types.GfSpQuotaInfo{}, err
+	}
+
+	return quotaResult, nil
+}
+
+// QuerySPHasEnoughQuotaForMigrateBucket is used to query src sp's bucket quota at approval phase
+func (s *GfSpClient) QuerySPHasEnoughQuotaForMigrateBucket(ctx context.Context, srcSPEndpoint string, bucketID uint64) error {
+	req, err := http.NewRequest(http.MethodGet, srcSPEndpoint+MigrateQueryBucketQuotaHasEnoughQuotaPath, nil)
+	if err != nil {
+		log.CtxErrorw(ctx, "client failed to connect to gateway", "endpoint", srcSPEndpoint, "error", err)
+		return err
+	}
+	req.Header.Add(GnfdBucketMigrateBucketIDMsgHeader, strconv.FormatUint(bucketID, 10))
+	resp, err := s.HTTPClient(ctx).Do(req)
+	if err != nil {
+		log.Errorw("failed to query migrate bucket quota msg", "bucket_id", bucketID, "error", err)
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("failed to query latest bucket quota, bucket(%d), status_code(%d), endpoint(%s)", bucketID, resp.StatusCode, srcSPEndpoint)
+	}
+
+	signedMsg := resp.Header.Get(GnfdSignedApprovalMsgHeader)
+
+	if signedMsg == "true" {
```

Comment:
it is not safe if just a "true",  may be use the private key to sign the approval,,